File: scripts/remove-bg.ts
/**
 * Removes the checkerboard background from the Firefly sprite sheet,
 * then crops into individual mascot frames with proper transparency.
 *
 * Usage: npx tsx scripts/remove-bg.ts
 */
import sharp from 'sharp'
import { resolve, dirname } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const root = resolve(__dirname, '..')
const src = resolve(root, 'dino_images/Firefly_Gemini Flash_1.png')

// Load raw pixels (no alpha on source, so we'll add it)
const { data: rawData, info } = await sharp(src)
  .ensureAlpha()
  .raw()
  .toBuffer({ resolveWithObject: true })

const { width, height, channels } = info // channels = 4 (RGBA)
const pixels = new Uint8ClampedArray(rawData)

// Flood-fill from image edges to find the checkerboard background.
// The checkerboard alternates between:
//   light grey: ~(191-195, 191-195, 190-193)
//   white:      ~(248-255, 248-255, 248-255)

function isBackground(r: number, g: number, b: number): boolean {
  // White squares
  if (r >= 240 && g >= 240 && b >= 240) return true
  // Light grey squares
  if (r >= 180 && r <= 205 && g >= 180 && g <= 205 && b >= 178 && b <= 205
      && Math.abs(r - g) < 10 && Math.abs(g - b) < 10) return true
  return false
}

const W = width
const H = height
const visited = new Uint8Array(W * H)
const isBg = new Uint8Array(W * H)
const queue: number[] = []

function seed(idx: number) {
  const pi = idx * channels
  if (!visited[idx] && isBackground(pixels[pi], pixels[pi + 1], pixels[pi + 2])) {
    visited[idx] = 1
    isBg[idx] = 1
    queue.push(idx)
  }
}

// Seed from all four edges
for (let x = 0; x < W; x++) { seed(x); seed((H - 1) * W + x) }
for (let y = 0; y < H; y++) { seed(y * W); seed(y * W + W - 1) }

// BFS flood fill
while (queue.length > 0) {
  const idx = queue.shift()!
  const x = idx % W, y = Math.floor(idx / W)
  for (const [dx, dy] of [[-1,0],[1,0],[0,-1],[0,1]]) {
    const nx = x + dx, ny = y + dy
    if (nx < 0 || nx >= W || ny < 0 || ny >= H) continue
    const ni = ny * W + nx
    if (visited[ni]) continue
    const pi = ni * channels
    if (isBackground(pixels[pi], pixels[pi + 1], pixels[pi + 2])) {
      visited[ni] = 1
      isBg[ni] = 1
      queue.push(ni)
    }
  }
}

// Apply: set background pixels to fully transparent
for (let i = 0; i < W * H; i++) {
  if (isBg[i]) pixels[i * channels + 3] = 0
}

// Reconstruct a full RGBA image
const fullRgba = await sharp(Buffer.from(pixels), { raw: { width: W, height: H, channels: 4 as const } })
  .png()
  .toBuffer()

// Now crop into quadrants and export
const hw = Math.floor(W / 2)
const hh = Math.floor(H / 2)

const frames = [
  { name: 'neutral',  left: 0,  top: 0  },
  { name: 'happy',    left: hw, top: 0  },
  { name: 'sad',      left: 0,  top: hh },
  { name: 'thinking', left: hw, top: hh },
]

console.log(`Source: ${W}×${H} → four ${hw}×${hh} frames`)

for (const frame of frames) {
  const webpOut = resolve(root, `public/mascot/${frame.name}.webp`)
  const pngOut  = resolve(root, `public/mascot/${frame.name}.png`)

  // Extract quadrant, trim transparency, save to intermediate buffer
  const trimmedBuf = await sharp(fullRgba)
    .extract({ left: frame.left, top: frame.top, width: hw, height: hh })
    .trim({ threshold: 5, background: { r: 0, g: 0, b: 0, alpha: 0 } })
    .png()
    .toBuffer()

  await sharp(trimmedBuf).resize({ width: 400, withoutEnlargement: true }).webp({ quality: 85, effort: 4 }).toFile(webpOut)
  await sharp(trimmedBuf).resize({ width: 400, withoutEnlargement: true }).png({ compressionLevel: 9 }).toFile(pngOut)

  const webpMeta = await sharp(webpOut).metadata()
  const pngMeta  = await sharp(pngOut).metadata()
  const wStat = (await import('fs')).statSync(webpOut)
  const pStat = (await import('fs')).statSync(pngOut)
  console.log(`  ${frame.name}: ${webpMeta.width}×${webpMeta.height}, webp ${Math.round(wStat.size/1024)}KB (${webpMeta.hasAlpha ? 'alpha ✓' : 'NO ALPHA'}), png ${Math.round(pStat.size/1024)}KB (${pngMeta.hasAlpha ? 'alpha ✓' : 'NO ALPHA'})`)
}

console.log('\nDone.')

File: scripts/remove-bg.py
"""
Removes checkerboard background from Firefly sprite sheet, crops into 4 frames.
Usage: python3 scripts/remove-bg.py
"""
from PIL import Image
import numpy as np
from collections import deque
import os

root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
src = os.path.join(root, 'dino_images', 'Firefly_Gemini Flash_1.png')

img = Image.open(src).convert('RGBA')
data = np.array(img)
H, W = data.shape[:2]

def is_background(r, g, b):
    # White squares
    if r >= 240 and g >= 240 and b >= 240:
        return True
    # Light grey squares (~191-195, neutral)
    if 178 <= r <= 210 and 178 <= g <= 210 and 178 <= b <= 210:
        if abs(int(r) - int(g)) < 15 and abs(int(g) - int(b)) < 15:
            return True
    return False

# Flood fill from all edges to identify background
bg_mask = np.zeros((H, W), dtype=bool)
visited = np.zeros((H, W), dtype=bool)
queue = deque()

def seed(y, x):
    r, g, b, _ = data[y, x]
    if not visited[y, x] and is_background(r, g, b):
        visited[y, x] = True
        bg_mask[y, x] = True
        queue.append((y, x))

hw_src, hh_src = W // 2, H // 2

for x in range(W):
    seed(0, x)
    seed(H - 1, x)
    # Also seed from centre horizontal line (guaranteed background between quadrants)
    seed(hh_src, x)
for y in range(H):
    seed(y, 0)
    seed(y, W - 1)
    # Also seed from centre vertical line (guaranteed background between quadrants)
    seed(y, hw_src)

while queue:
    y, x = queue.popleft()
    for dy, dx in [(-1,0),(1,0),(0,-1),(0,1)]:
        ny, nx = y + dy, x + dx
        if 0 <= ny < H and 0 <= nx < W and not visited[ny, nx]:
            r, g, b, _ = data[ny, nx]
            if is_background(r, g, b):
                visited[ny, nx] = True
                bg_mask[ny, nx] = True
                queue.append((ny, nx))

# Apply mask: set background alpha to 0
data[bg_mask, 3] = 0

full_img = Image.fromarray(data)

print(f'Source: {W}×{H}  Background pixels removed: {bg_mask.sum()}')

# Quadrant crop + trim
hw, hh = W // 2, H // 2
frames = [
    ('neutral',  0,  0),
    ('happy',   hw,  0),
    ('sad',      0, hh),
    ('thinking',hw, hh),
]

out_dir = os.path.join(root, 'public', 'mascot')

for name, left, top in frames:
    quad = full_img.crop((left, top, left + hw, top + hh))

    # Trim transparent edges
    bbox = quad.getbbox()
    if bbox:
        quad = quad.crop(bbox)

    # Resize to max 400px wide
    if quad.width > 400:
        ratio = 400 / quad.width
        quad = quad.resize((400, int(quad.height * ratio)), Image.LANCZOS)

    png_path = os.path.join(out_dir, f'{name}.png')
    quad.save(png_path, 'PNG', optimize=True)

    # WebP with alpha
    webp_path = os.path.join(out_dir, f'{name}.webp')
    quad.save(webp_path, 'WebP', quality=85, method=4)

    png_kb = os.path.getsize(png_path) // 1024
    webp_kb = os.path.getsize(webp_path) // 1024
    mode = Image.open(webp_path).mode
    print(f'  {name}: {quad.width}×{quad.height}  webp {webp_kb}KB ({mode})  png {png_kb}KB')

print('\nDone.')
